cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -951,7 +951,7 @@ class ClientBuilder {
       client_config_.caPath = ToAwsString(internal::global_options.tls_ca_dir_path);
     }
 
-    const bool use_virtual_addressing = options_.endpoint_override.empty();
+    const bool use_virtual_addressing = options_.use_virtual_addressing;
```

Comment:
This will break the current behavior where, if `endpoint_override` is set, virtual addressing is disabled.

Perhaps we should make `use_virtual_addressing` a `std::optional<bool>` instead?

Or perhaps we could call it `force_virtual_addressing` and have the following logic:
```c++
    const bool use_virtual_addressing = options_.endpoint_override.empty() || options_.force_virtual_addressing;
```

Reply:
ok, I have changed the code as the last logic